CondFormats/SiPixelObjects/src/SiPixelFEDChannelContainer.cc

```diff
@@ -0,0 +1,125 @@
+  if( m_scenarioMap.find(theScenarioId) != m_scenarioMap.end()){
+    edm::LogWarning("SiPixelFEDChannelContainer") << "Scenario: " <<theScenarioId <<" is already in the map!"<<std::endl;
+    return;
+  } else {
+    m_scenarioMap.emplace(theScenarioId,theBadFEDChannels);
+  }
+}
+
+//****************************************************************************//
+SiPixelFEDChannelContainer::SiPixelFEDChannelCollection SiPixelFEDChannelContainer::getSiPixelBadFedChannels(const std::string &theScenarioId) const {
+  SiPixelBadFEDChannelsScenarioMap::const_iterator it = m_scenarioMap.find(theScenarioId);
+
+  if (it != m_scenarioMap.end()){
+    return it->second;
+  } else {
+    throw cms::Exception("SiPixelFEDChannelContainer")<< "No Bad Pixel FEDChannels defined for Scenario id: " << theScenarioId << "\n";
+  }
+}
+
+//****************************************************************************//
+SiPixelFEDChannelContainer::SiPixelFEDChannelCollection& SiPixelFEDChannelContainer::getSiPixelBadFedChannels(const std::string &theScenarioId) {
+  if(m_scenarioMap.find(theScenarioId) != m_scenarioMap.end()){
+    return m_scenarioMap[theScenarioId];
+  } else {
+    throw cms::Exception("SiPixelFEDChannelContainer")<< "No Bad Pixel FEDChannels defined for Scenario id: " << theScenarioId << "\n";
+  }
+}
+
+//****************************************************************************//
+std::vector<PixelFEDChannel>& SiPixelFEDChannelContainer::getSiPixelBadFedChannelsInDetId(const std::string &theScenarioId,DetId theDetId) {
+
+  if(m_scenarioMap.find(theScenarioId) == m_scenarioMap.end()){
+    throw cms::Exception("SiPixelFEDChannelContainer")<< "No Bad Pixel FEDChannels defined for Scenario id: " << theScenarioId << "\n";
+  } else {
+    if(m_scenarioMap[theScenarioId].find(theDetId) == m_scenarioMap[theScenarioId].end()){
+      throw cms::Exception("SiPixelFEDChannelContainer")<< "No Bad Pixel FEDChannels defined for DetId:" <<theDetId <<" in Scenario id: " << theScenarioId << "\n";
+    }
+    return (m_scenarioMap[theScenarioId])[theDetId];
+  }
+}
+
+//****************************************************************************//
+std::unique_ptr<PixelFEDChannelCollection> SiPixelFEDChannelContainer::getDetSetBadPixelFedChannels(const std::string &theScenarioId) const {
+
+  if(m_scenarioMap.find(theScenarioId) == m_scenarioMap.end()){
+    throw cms::Exception("SiPixelFEDChannelContainer")<< "No Bad Pixel FEDChannels defined for Scenario id: " << theScenarioId << "\n";
+  }
+
+  std::unique_ptr<PixelFEDChannelCollection> disabled_channelcollection = std::make_unique<edmNew::DetSetVector<PixelFEDChannel> >();
+  auto SiPixelBadFedChannels = m_scenarioMap.at(theScenarioId);
+  for(const auto &entry : SiPixelBadFedChannels){
+    disabled_channelcollection->insert(entry.first, entry.second.data(), entry.second.size());
+  }
+  return disabled_channelcollection;
```

Comment:
HI @makortel, I am sorry I am not sure to understand. 
What would it be the proposition on

> about do the translation above once

?
returning by const reference can be done even in the current implementation (and change the function call in the digitizer).
One in principle can also use the other method which returns `SiPixelFEDChannelContainer::SiPixelFEDChannelCollection& `
and wrap the `DetSetVector` in the client code (but we wanted to provide a common interface for the various clients.

Reply:
@mmusich Let's say that I don't like unnecessarily repeated computations, and I think new code should try to minimize that.

Reply:
@makortel if we can avoid unnecessary computations I am all for it, I just don't understand technically how you propose to avoid it.

Reply:
One option would be to `SiPixelFEDChannelContainer` itself to store e.g. `std::unordered_map<std::string, PixelFEDChannelCollection>` (it looks to me that currently no-one is calling `getSiPixelBadFedChannels()`).

Another option would be to add another layer of ESProduct+ESProducer that takes `SiPixelFEDChannelContainer` and does the translation from `SiPixelFEDChannelCollection` to `PixelFEDChannelCollection` once (per scenario) in its constructor. The clients would then use this new ESProduct to get PixelFEDChannelCollection.

Reply:
thanks for the hints.
About option 1):

>  store std::unordered_map<std::string, PixelFEDChannelCollection> 

this is actually the first thing I tried, but unfortunately as 
https://github.com/cms-sw/cmssw/blob/02d4198c0b6615287fd88e9a8ff650aea994412e/DataFormats/SiPixelDetId/interface/PixelFEDChannel.h#L15

involves a `edmNew::DetSetVector` (which is a non-trivial object) I would need to code (rather compliacted) dedicated boost serialization rules (without any guarantee not to break schema in future, if someone ever changes the `DetSetVector` class). see similar concern from Giacomo @govi [here](https://github.com/cms-sw/cmssw/pull/25466#discussion_r241013375)

About option 2):

> add another layer of ESProduct+ESProducer that takes `SiPixelFEDChannelContainer` and does the translation from `SiPixelFEDChannelCollection` to `PixelFEDChannelCollection` once (per scenario) 

OK, this is something I can explore.

Reply:
Third option would be a helper function that takes `SiPixelFEDChannelContainer` as an input and produces `PixelFEDChannelCollection`, and then client calls that function if an `ESWatcher` tells that the IOV has changed. That would be simpler from the ESProducer side, but would lead to more complex client code and the translation to be done more times than with the option 2.

Reply:
> tells that the IOV has changed. 

but the IOV will never change in simulation (i.e. there is only one payload which contains all possible scenarios of FED Channels to kill)

Reply:
CMS is working on the possibility of run dependent MC. In that case IOVs will change during a job.

Reply:
In case CMS provide real run-depedent MC, this whole setup becomes completely unnecessary. Just we populate the `SiPixelQuality` with the same tags we have in data.